### PR TITLE
Style guide

### PR DESCRIPTION
**Pull request by Nihlus**
### Purpose of this PR

* This PR adds a style guide for the project, containing a set of rules that describe how code should and shouldn't be formatted and written.

At the moment, the guide is not complete, and feedback is very much wanted. Please have a read through the [draft](https://github.com/Nihlus/opentk/blob/style-guide/documentation/style-guide/main.pdf) and give your thoughts.

### CODE REVIEW

**Comment by FreezyLemon** [comment]
We've talked about it before on discord, but here are some small things I wanted to mention:

- 3.2.10: There's _some_ mistake here. If all tuple members are considered properties, shouldn't they be PascalCase? (see 3.2.8 and 3.2.9)
- 4.0.1: Why the "all static declarations must come last in their section"? i think i've seen the static properties put first more often than the other way around..

There are also some listings with incorrect indentation (8 spaces instead of 4):
Listing 6, 7, 16, 19 (or rather, the example below Listing 18).

Oh, and very big 👍 for 6.3. Maybe even split that out into 2 separate rules:
1. `null` should be treated as an exceptional value
2. null checks should be performed using pattern matching

IMO no. 1 can sometimes require a bit of a different approach when implementing functionality, so it might be a good idea to have it as its own rule instead of just having it as a small part in another rule.

**Comment by VPeruS** [comment]
Can you rename "The OpenTK Style Guide" to "StyleGuide" or "style-guide"?

File: documentation/The OpenTK Style Guide/main.tex [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+
+\begin{must}{Files must have the correct header}
+Files shall contain a file header, containing the following text.
+\end{must}
+
+\begin{code}[File header]
+// 
+//  <filename>
+//
+//  Copyright (C) <year of creation> OpenTK
```
Comment (VPeruS) [inline documentation/The OpenTK Style Guide/main.tex]: Previous authors cannot be changed or removed, they should be mentioned.

Comment (Nihlus) [reply documentation/The OpenTK Style Guide/main.tex]: Good point. If the file has listed authors, they should be included.

**Comment by Nihlus** [comment]
@VPeruS Any particular reason?

@FreezyLemon 
* 3.2.10 - you're absolutely right, my bad.
* 4.0.1 - in general, the idea was to put the working code of the class itself first so that static methods get a lower mental priority. Statics are overused in the library, and placing them after the instance methods (where the bulk of the work is to be done) assists in compartmentalizing that. It's something we can change, of course, if that doesn't make sense to anyone else :P

* 6.3 - yeah, sounds like a plan!

**Comment by VPeruS** [comment]
@Nihlus spaces must be excluded while using command line interpeter as \  or " "

**Comment by FreezyLemon** [comment]
Hm... 4.0.1 does sound reasonable. I'm not against it _that_ much, it just didn't seem to make too much sense to me.

I don't think I mind either way, honestly. As long as it's uniform across the project

**Comment by Nihlus** [comment]
@VPeruS It's not really a folder that the command line will touch all that much.

**Comment by Nihlus** [comment]
@VPeruS Ah, hell, you're right, sorry. It simplifies things.

**Comment by varon** [comment]
This is really an amazing document, @Nihlus. You should be super proud of the work you've done on it; including the LaTeX source is a great touch too.

Only thing I'm a little worried about is that I fear this might deter new contributors as it's both a lot to read and a huge number of rules to keep in your head at once while programming.

Is it possible to configure an automated tool to just format everything and format using that?

**Comment by Nihlus** [comment]
The majority of the rules are either covered by sane defaults in your IDE, or by using a feature-rich editor with support for loading the configurations stored in the repo (VS with R#, Rider, emacs, VSCode (at least partially), etc). StyleCop certainly helps, too.

As a document, its main purpose is for us to have something to point to when reviewing to be able to say "hey, this is what's wrong and we have a rule for it" - reviews get less arbitrary to both us and contributors.

We miiight be able to spend some time writing our own Roslyn analyzer to check for the rules. It's certainly possible, but someone has to find the time to actually do it.

**Comment by Nihlus** [comment]
@varon Thoughts on having this in a separate repo from the main OpenTK repo? It's related, but not strictly speaking part of the library.

**Comment by varon** [comment]
I don't see any point in a separate repo for this.

For the rules themselves:

4.1 - Rename to "Methods should use the least permissive access modifier possible". 

There's also the internal modifier, which can help for library-only code.

6.1 - An explicit type can be used to force a coercion at some points, that would also constitute an acceptable use.

I'm a touch hesitant about 6.3.

I'd argue that `value == null` is the null check that 99% of C# developers would write in any normal circumstance. The not-null check is even more awkward, requiring extra brackets.

Even from an F# perspective, where pattern matching has been around for a long time, we still tend to use `if then else` statements for null checking rather than pattern matching against it.

For two new suggestions:
We should note that it's not part of our style to DEallocate memory.
Additionally, performance is a priority and code should be well optimized and fast.

**Comment by Nihlus** [comment]
My general thoughts about 6.3 can be read here:
http://sharkman.asuscomm.com/blog/is-null-versus-null-in-c/

One highlight is that it's actually faster, and uses fewer instructions.

Performance being a priority, I feel, might not warrant a rule - then
again, I don't see a compelling reason to not make it one.

As for deallocating memory, could you give me an example? Typically, the
deallocation rule just follows ownership rules - if you own the memory, you
are responsible for freeing it.

**Comment by varon** [comment]
Yeah - I'm familiar with how it works internally. But it's a bit like requiring people call `Object.ReferenceEquals(value, null)` in earlier versions of C# - sometimes correct but very unusual to make a rule out of.

Aside from the IL comparison, do you have any benchmarks? What's the .Net standard or core team's rulings on this?

On deallocations, I mean only in a managed context - if you instantiate an array, either keep that array around forever or take it as a parameter and make the method void.

OpenTK is typically used in soft-realtime applications which need to limit or avoid GC as much as possible. Allocations themselves aren't a huge problem, as the system will very quickly realize "oh hey, this memory isn't getting GCd" and adjust the collection threshold internally.

The intent is really to stop people doing silly linq or working with tons of intermediate objects.

**Comment by Nihlus** [comment]
@varon I don't have any benchmarks at this time - however, the real-life implementations of equality overloads are going to vary wildly. The rule is there for two primary reasons, which I believe are extremely important - correctness and exceptionality. When we do a null check, we're wanting to see if the object is null - not if it behaves like a null, or is equivalent in certain cases to a null. The semantics of code becomes very different when the equality operator is overloaded against null, and its use is pure evil.

By guarding ourselves like this, we not only make developers more aware of the issue, but we also completely eliminate nasty surprises from cases like this.

The .NET team does not make any reference to similar rules. The times I've brought it up, however, it's been well received.

Deallocations sounds like a good thing to make a note of. OpenTK doesn't consider itself the owner of memory it returns, and methods that should return an array should provide a ref overload.

**Comment by FreezyLemon** [comment]
Some questions about 2.1 ("Line width must not exceed 120 columns"):
1. What does the description mean, exactly? Should the line be broken at the nearest word boundary _inwards_ (so the line **always** stays shorter than 120 characters) or _outwards_ (so it can technically be more than 120 characters, in which case, the rule name is kind of wrong)?
2. Is there a specific reason for calling it line "width" and "column"? I think that the term "line length" in conjunction with "character" is a lot more common.
3. About 2.1.4: How about renaming "newline" character to "line feed" character here? The term "newline" is kind of ambiguous (maybe not in theory, but definitely in practice: most people use newline for the combination of characters that defines the end of a line; at least in my experience), and i think more people would understand "line feed" more easily (from 'CRLF' and 'LF'), especially since you call out the absence of a CR explicitly.

**Comment by Nihlus** [comment]
@FreezyLemon 

1. Inwards. The line should always stay shorter than 120 columns.
2. A single character may occupy less than one column in a text document. Typically, this won't be the case or an issue for us, but it's still something to be clear about. It's not about 120 characters in a row, but rather that the characters on one line should not span more than 120 columns.
3. Good catch, I agree.

**Comment by DylanFPS** [comment]
Really just a nitpick, but it kinda gets to me that you use "Listing" instead of something more generic like "Figure". Again - this doesn't really affect anything much, but just a personal preference.

**Comment by Nihlus** [comment]
@DylanFPS That's not up to me - rather, it is how LaTeX generates them.

**Comment by FreezyLemon** [comment]
Reminder (from Discord) to add a rule for abbreviations:
- "Official" Microsoft way: only fully capitalize 2-letter abbreviations (ID, PR; but Url, Led)
vs
- other options, like just using abbreviations normally (API, URL, Id, etc.)

**Comment by VPeruS** [comment]
Voting for Id, Url, Led, Pr

**Comment by varon** [comment]
Let's figure out whatever scheme we use and just make it a rule.

**Comment by DylanFPS** [comment]
I agree with @VPeruS to be honest. Id > ID, Uri > URI, etc.

**Comment by Nihlus** [comment]
Previously discussed alterations have been implemented. Any other thoughts, or should we merge it in?

**Comment by Jijjy** [comment]
I'm lurking, but one convention i've personally followed for a while is putting operands and punctuation on the left, i.e.

```
var foo = SomeEnum.A
    | SomeEnum.B
    | SomeEnum.C
    | SomeEnum.E;

var bar = IsVisible
    && IsEnabled
    && SomethingElse;

var vm = new ViewModel
{
    foo = "bar"
    , zero = 0
    , pi = 3.0
};
```

It doesn't seem widely used, I just like seeing things aligned. One example where this can help:

```
TooManyArgs(1,2,3,4,5,6,7,8,9, foo, bar,
    GetSomething());
```

`GetSomething` at first glance might look separate whereas:

```
TooManyArgs(1,2,3,4,5,6,7,8,9, foo, bar
    , GetSomething());
```

makes it a little more obvious that the result of `GetSomething` is being passed to `TooManyArgs`.

Or:

```
var bar = IsVisible
    && IsEnabled
    && Lorem
    && Ipsum
    || SomethingElse;
```

Helps you see right away `|| SomethingElse` might be a mistake, easier than reading:

```
var bar = IsVisible &&
    IsEnabled &&
    Lorem &&
    Ipsum ||
    SomethingElse;
```

**Comment by Nihlus** [comment]
@Jijjy I like the alignment for operators, but not for punctuation. That looks messy and like a mistake - on top of that, it's covered by rules 2.2.6 and 2.2.7.

I'd be for including the operator alignment as a rule. Commas, no.

**Comment by acid-chicken** [comment]
How about enhancing EditorConfig?
FYI: https://docs.microsoft.com/visualstudio/ide/editorconfig-code-style-settings-reference

**Comment by varon** [comment]
Whichever rules we go with, just double check that they can actually be done by the current auto-format tools.

**Comment by Nihlus** [comment]
@varon A number of the rules in this document are there for the sole reason that autoformatting can't catch them - @FreezyLemon was looking into building us a custom Roslyn analyzer for OpenTK, as I recall.

**Comment by varon** [comment]
I'd advise making those recommendations rather than requirements; it's important we're not rejecting otherwise-good PRs for stylistic reasons.

**Comment by FreezyLemon** [comment]
@acid-chicken IIRC, editorconfig doesn't even begin to cover the style rules we want to apply. It's a nice idea and I've taken a look at it too, but it just doesn't seem to be enough for OpenTK's ruleset.

Yes, I'm still looking into writing a Roslyn analyzer - the tree structure takes some getting used to, it's not very intuitive to me.

Custom analyzers **can** contain auto-fixes - that is why I think there should be an ecosystem (e.g. with said custom analyzer) to catch all and auto-fix most of these style "errors" so that the little manpower we have is not wasted on manual code style reviews..

Consistent code styling + automated style checks & fixes = having the cake and eating it too, IMO

all that's left is actually writing the darn thing...

**Comment by varon** [comment]
>In general, it'll be a quick fix for any prospective PR to correct any style issues they might have, and the author should have read the guide beforehand - or, at the very least, understand that it exists and we have it for a reason.

The main thing I want to ensure is that we make it as easy to contribute as possible.  Being forced to read and apply a 20+ page style guide is often way more effort than making contribution itself.

I think the difference of opinion comes down to this: I'd argue that having more PRs is more useful than having perfectly formatted code. Most of us can read just about anything in any format, it's not like it makes much of a difference.

Clean style is very nice, but it's something the maintainers and significant contributors should be aware of, not mom-and-pop PRs of a few hundred lines.

>Consistent code styling + automated style checks & fixes = having the cake and eating it too, IMO

Instead of trying to write tooling to our standards, can't we just change the coding style to minimize the effort required to get there? i.e. "default VS/Rider C# formatting".
I don't think the exact formatting choices really matter as long as they're consistent, and above all EASY to apply.

**Comment by Nihlus** [comment]
Practically all of our set rules and standards are the standard way of doing things. We deviate in very few (if any) places. Anything that isn't standard fills in a gap where there is no standard way.

Those mom and pop PRs is exactly where it'll be quick and easy to point out and apply these fixes before they accumulate in the codebase. I want to reduce future costs by making sure things are correct and won't need another cleanup pass in the future the moment it gets merged.

**Comment by Nihlus** [comment]
Or to sum up - it's better to have many small PRs and a way to explain what needs to be fixed in them, than have many small PRs and have to do it yourself later.

**Comment by varon** [comment]
Seeing as we've already spent the effort writing the guide, I can't argue with that.

**Comment by peppy** [comment]
I hope all of these are being enforced using r#/inspectcode + `.DotSettings` or roslyn inspectors. Generally all stylistic changes should be applicable via resharper's "format code", taking any manual labour out of the process. Also presents to the developer in a format they can understand (IDE hinted errors/warnings) instead of attempting to parse a lengthy pdf.

**Comment by Nihlus** [comment]
All rules that have an applicable automated inspection will be configured
to match, yes. There are some that lack inspections.

**Comment by peppy** [comment]
Awesome.

On top of the exposed inspections, resharper can do a lot more in terms of enforcing code styling via it's `format document` functions. Can be a good way to tidy up PRs which don't confirm with a simple automated operation (and it can be run solution-wide to keep things in-line too). These rules can be part of `DotSettings` even though they aren't inspections themselves, would recommend taking a look!

**Comment by varon** [comment]
@Nihlus/ @DylanFPS. 
It would be great if we could export this thing as HTML too and get it on the website. Would make it pretty easy to reference and refer to.

**Comment by varon** [comment]
Also, do we have a final draft? Would be good to merge this.

**Comment by DylanFPS** [comment]
I’m not to familiar with stylecop, but I could probably add it to the site using pdf.js, or embed it or something

**Comment by Nihlus** [comment]
I'll see if I can export it for web - should be doable. Embedding a PDF might be more reliable, though. 

Final draft - yep, I'm happy with it as is. I would like a joint decision on the operator alignment rule before it's all green.

**Comment by varon** [comment]
That's really great work! Thanks!

**Pull request merged**